Feat 127/scheduling machines

Talos10 opened the pull request:
# Description

Modified the database and created the routes in order to be able to create the machines which will handle the orders.

Issue reference: #127 

## Type of change

- [ ] BUG (Fixing or documenting a bug)
- [X] FEAT (Add a new feature to the project)
- [ ] CHORE (Something that needs to be done outside of the issue)
- [ ] DOCS (Implement documentation)
- [ ] TASK (A task that must be done to complete an issue)

# Checklist:

- [X] My pull request references the issue in the comments
- [X] My code follows the style guidelines of this project
- [X] I have performed a self-review of my own code
- [X] I have made corresponding changes to the documentation
- [X] My changes generate no new warnings
- [X] I have added tests that prove my fix is effective or that my feature works
- [X] New and existing unit tests pass locally with my changes

Talos10 committed:
FEAT-127: Modified db and added backend routes for the creation of machines.

Talos10 committed:
FEAT-127: Added machine tests.

Talos10 committed:
FEAT-127: Removed comment in sql script.

phong1233 commented on backend/src/Machine/machine.service.ts:
```diff
@@ -0,0 +1,55 @@
+        return allMachines;
+    }
+
+    public async createNewMachine(status: string, numberOrderCompleted: number): Promise<number> {
```
i think create machine should not take any param and default to 'free' and 0

Talos10 replied:
Done!

phong1233 reviewed (commented):
Looks good! Left a comment. Also would it be possible to add comments for the service and the model as well?

Talos10 committed:
Added commits to the service and model files. Also made it so that creating a machine can be done without any parameters.

phong1233 reviewed (approved):
To the moon 🚀🚀🚀💎🙌💎🙌💎🙌💎🙌

N0ot-No0t commented:
![image](https://user-images.githubusercontent.com/27740712/112354784-665f1f80-8ca3-11eb-8ce2-facd882bb576.png)

I don't know if it's a good idea to have an endpoint that allows the deletion of ALL machines in the DB. I feel like there's some security risks here.

Talos10 commented:
> ![image](https://user-images.githubusercontent.com/27740712/112354784-665f1f80-8ca3-11eb-8ce2-facd882bb576.png)
> 
> I don't know if it's a good idea to have an endpoint that allows the deletion of ALL machines in the DB. I feel like there's some security risks here.

This is more for us in case we want to delete all the machines for some reason. The db also checks if a machine is currently processing an order or not and so if 1 machine out of 10 is busy with an order, no machines get deleted. I don't get where the security risk is therefore. Can't someone just his the "delete one machine" API end point 10 times instead?

sebastien-blain commented on backend/src/Machine/machine.controller.ts:
```diff
@@ -0,0 +1,134 @@
+            passport.authenticate('jwt', { session: false }),
+            async (req: Request, res: Response) => {
+                const id = Number(req.params.machineId);
+                const { status, numberOrderCompleted } = req.body;
```
Why do you let the client decide on the number of orders completed instead of adding 1 to the current number of orders completed?

Why don't you abstract the concept of updating a machine by creating routes to either start or complete an order? That way the client only has to call either start or complete with a machine id.

N0ot-No0t commented:
> > ![image](https://user-images.githubusercontent.com/27740712/112354784-665f1f80-8ca3-11eb-8ce2-facd882bb576.png)
> > I don't know if it's a good idea to have an endpoint that allows the deletion of ALL machines in the DB. I feel like there's some security risks here.
> 
> This is more for us in case we want to delete all the machines for some reason. The db also checks if a machine is currently processing an order or not and so if 1 machine out of 10 is busy with an order, no machines get deleted. I don't get where the security risk is therefore. Can't someone just his the "delete one machine" API end point 10 times instead?

If it only deletes non-busy machines then I guess it's alright. I just thought if someone would make a mistake in the endpoint, it could be problematic. If anyone would've really wanted to delete all of the machines for some reason then they would have to call the endpoint to delete singular machines instead (less risk to delete the entire table). I also noticed the same architecture for the other services and models in the backend (for example, the users). I don't think we should keep these endpoints in our code and give the possibility to delete the entirety of the content of a table with just one request. I highly doubt companies like Twitter or Facebook allow the possibility to delete their population of users with one request.

phong1233 reviewed (commented):
To address the comments, I think we should only keep 4 endpoints
- Get all machine
- Get machine by id
- Get machine by status
- Create machine

I don't think we will implement delete or update in the ui but you can keep the methods in the service since I will be using them for scheduling

sebastien-blain commented:
> To address the comments, I think we should only keep 4 endpoints
> 
> * Get all machine
> * Get machine by id
> * Get machine by status
> * Create machine
> 
> I don't think we will implement delete or update in the ui

Without knowing the implementation details (how the machines will be used in the scheduling backend), I agree with this suggestion.

N0ot-No0t commented:
> To address the comments, I think we should only keep 4 endpoints
> 
> * Get all machine
> * Get machine by id
> * Get machine by status
> * Create machine
> 
> I don't think we will implement delete or update in the ui

We should perhaps also remove endpoints related to deletion from other models and services like the users.

Talos10 committed:
Merge branch 'main' into FEAT-127/scheduling_machines

Talos10 committed:
FEAT-127: Removed the update and delete endpoints and added an additional endpoint in order to be able to get the machines by status.

Talos10 commented:
Addressed your comments by removing the update and delete endpoints and by adding an additional endpoint in order to be able to get the machines by status.

N0ot-No0t commented:
> Addressed your comments by removing the update and delete endpoints and by adding an additional endpoint in order to be able to get the machines by status.

I see you deleted the endpoint on the controller, but isn't it going to create dead code in the service and model if you leave the "delete all machines" endpoint there?

phong1233 commented on backend/test/machineTests/service.test.ts:
```diff
@@ -43,29 +50,4 @@ describe('Machine Service Test', () => {
         const res = await machineService.findMachineById(1);
         expect(res).to.equal('foo');
     });
-
```
I think you can put back these tests unless they depend on the controller

Talos10 committed:
FEAT-127: Readded some tests and removed the delete methods from the service and model.

Talos10 committed:
FEAT-127: Deleted obsolete tests.
